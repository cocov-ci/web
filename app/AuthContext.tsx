'use client'

import { deleteCookie, setCookie } from 'cookies-next'
import { useRouter } from 'next/navigation'
import React, {
  createContext,
  useContext,
  useEffect,
  useMemo,
  useState,
} from 'react'

import { fetchAPI } from 'services/api/fetch'
import { fetchUser } from 'services/api/users'
import { UserProps } from 'types/User'

type AuthState = {
  loading: boolean
  login: () => void
  logout: () => void
  user: UserProps
}

const StateContext = createContext<AuthState>({
  user: null,
  login: () => null,
  logout: () => null,
  loading: false,
})

export const AuthProvider = ({ children }: { children: React.ReactNode }) => {
  const [user, setUser] = useState<UserProps>(null)
  const [loading, setLoading] = useState<boolean>(true)
  const router = useRouter()

  useEffect(() => {
    const fetchData = async () => await fetchUser()

    fetchData()
      .then((data: UserProps) => setUser(data))
      .catch(error => {
        console.error(error)
      })
      .finally(() => setLoading(false))
  }, [])

  const login = async () => {
    setLoading(true)

    await fetchAPI(
      'https://run.mocky.io/v3/d6ac91ac-6dab-4ff0-a08e-9348d7deed51',
    )
      .then((data: UserProps) => {
        // TODO: IMPLEMENT SESSIONS
        setCookie('cocov_auth_token', data?.id)

        router.push('/')

        setUser(data)
      })
      .catch(error => {
        console.error(error)
      })
      .finally(() => setLoading(false))
  }

  const logout = () => {
    deleteCookie('cocov_auth_token')
    setUser(null)
    router.push('/')
  }

  const memoedValue = useMemo(
    () => ({
      user,
      loading,
      login,
      logout,
    }),
    [user, loading],
  )

  return (
    <StateContext.Provider value={memoedValue}>
      {children}
    </StateContext.Provider>
  )
}

export const useAuth = () => useContext(StateContext)
