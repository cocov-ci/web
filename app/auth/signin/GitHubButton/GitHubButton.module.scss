@import "styles/variables";

.gitHubButton {
  .button {
    width: 100%;
    position: relative;
    padding: 12px 16px;
    border-radius: 8px;
    cursor: pointer;
    background-color: $colorGrayDark;
    font-size: 17px;
    color: $colorWhite;
    border: none;

    img {
      position: absolute;
      top: 50%;
      transform: translateY(-50%);
      left: 15px;
    }
  }
}
