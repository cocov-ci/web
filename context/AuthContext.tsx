'use client'

import { useRouter, useSearchParams } from 'next/navigation'
import React, { createContext, useMemo, useState } from 'react'

import Auth from 'services/auth'
import { AuthBeginReponseProps, AuthPropsContext } from 'types/Auth'

export const AuthContext = createContext<AuthPropsContext>({
  login: () => null,
  logout: () => null,
  loading: false,
  isAuthenticated: false,
})

const SigninPage = '/auth/signin'

const AuthProvider = ({ children }: { children: React.ReactNode }) => {
  const [loading, setLoading] = useState<boolean>(false)
  const user = Auth.getAccount()
  const router = useRouter()
  const searchParams = useSearchParams()

  const login = async () => {
    setLoading(true)

    try {
      const data: AuthBeginReponseProps = await Auth.begin()

      window.location.href = searchParams.get('next')
        ? `${data.redirect_to}?next=${searchParams.get('next')}`
        : data.redirect_to
    } catch {
      setLoading(false)
      // TODO: HANDLE WITH THIS RESPONSE
    }
  }

  const logout = () => {
    Auth.deleteToken()
    Auth.deleteAccount()
    router.push(SigninPage)
  }

  const memoizedValue = useMemo(
    () => ({
      loading,
      login,
      logout,
      isAuthenticated: Boolean(user),
      user: user && { ...user },
    }),
    [loading, user],
  )

  return (
    <AuthContext.Provider value={memoizedValue}>
      {children}
    </AuthContext.Provider>
  )
}

export default AuthProvider
