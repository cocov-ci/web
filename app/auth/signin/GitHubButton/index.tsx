'use client'

import Image from 'next/image'

import { useAuth } from 'app/AuthContext'
import Button from 'app/common/Button'
import Loading from 'app/loading'

import styles from './GitHubButton.module.scss'

const GitHubButton = () => {
  const { login, loading } = useAuth()

  if (loading) return <Loading />

  return (
    <Button className={styles.button} onClick={(): void => login()}>
      <Image
        alt="GitHub logo"
        className={styles.logo}
        height="20"
        src="/icons/github.svg"
        width="20"
      />
      Log in with GitHub
    </Button>
  )
}

export default GitHubButton
