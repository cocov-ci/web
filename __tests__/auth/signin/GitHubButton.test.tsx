import { fireEvent, render, screen } from '@testing-library/react'

import AuthContext from '__tests__/__mocks__/AuthContext'
import GitHubButton from 'app/auth/signin/GitHubButton'

import '@testing-library/jest-dom'

const props = {
  login: jest.fn(),
  logout: jest.fn(),
  loading: false,
  isAuthenticated: false,
}

describe('app/auth/signin/GitHubButton', () => {
  it('renders GitHubButton', () => {
    render(
      <AuthContext {...props}>
        <GitHubButton />
      </AuthContext>,
    )

    expect(screen.getByRole('button')).toBeVisible()
  })

  it('triggers onClick event', () => {
    render(
      <AuthContext {...props}>
        <GitHubButton />
      </AuthContext>,
    )

    fireEvent.click(screen.getByRole('button'))
    expect(props.login).toHaveBeenCalledTimes(1)
  })

  it('renders GitHub button with "loading" param', () => {
    const { container } = render(
      <AuthContext {...props} loading={true}>
        <GitHubButton />
      </AuthContext>,
    )

    expect(screen.queryByRole('button')).toBeNull()
    expect(container.querySelector('img')).toBeVisible()
  })

  it('renders GitHubButton snapshots', () => {
    const { container } = render(
      <AuthContext {...props}>
        <GitHubButton />
      </AuthContext>,
    )

    expect(container).toMatchSnapshot()
  })
})
