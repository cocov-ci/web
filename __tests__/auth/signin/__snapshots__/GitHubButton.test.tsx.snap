// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/auth/signin/GitHubButton renders GitHubButton snapshots 1`] = `
<div>
  <div
    class="gitHubButton"
  >
    <button
      class="className button button secondary"
    >
      <img
        alt="GitHub logo"
        class="logo"
        data-nimg="1"
        decoding="async"
        height="20"
        loading="lazy"
        src="/icons/github.svg"
        srcset="/icons/github.svg 1x, /icons/github.svg 2x"
        style="color: transparent;"
        width="20"
      />
      Log in with GitHub
    </button>
  </div>
</div>
`;
